src/services/ProveGatewayService.ts

```diff
@@ -0,0 +1,181 @@
+
+const Mosaic = require('@openst/mosaic.js');
+
+export default class ProveGatewayService extends Observer<AuxiliaryChain> {
+  private gatewayRepository: GatewayRepository;
+
+  private messageRepository: MessageRepository;
+
+  private originWeb3: object;
+
+  private auxiliaryWeb3: object;
+
+  private auxiliaryWorkerAddress: string;
+
+  private gatewayAddress: string;
+
+  private auxiliaryChainId: number;
+
+  /**
+   *  Constructor
+   *
+   * @param gatewayRepository Instance of auxiliary chain repository.
+   * @param messageRepository Instance of message repository.
+   * @param originWeb3 Origin Web3 instance.
+   * @param auxiliaryWeb3 Auxiliary Web3 instance.
+   * @param auxiliaryWorkerAddress auxiliary worker address, this should be
+   *                               unlocked and added in web3 wallet.
+   * @param gatewayAddress Address of gateway contract on origin chain.
+   * @param auxiliaryChainId Auxiliary chain Id.
+   */
+  public constructor(
+    gatewayRepository: GatewayRepository,
+    messageRepository: MessageRepository,
+    originWeb3: object,
+    auxiliaryWeb3: object,
+    auxiliaryWorkerAddress: string,
+    gatewayAddress: string,
+    auxiliaryChainId: number,
+  ) {
+    super();
+
+    this.gatewayRepository = gatewayRepository;
+    this.originWeb3 = originWeb3;
+    this.auxiliaryWeb3 = auxiliaryWeb3;
+    this.auxiliaryWorkerAddress = auxiliaryWorkerAddress;
+    this.gatewayAddress = gatewayAddress;
+    this.messageRepository = messageRepository;
+    this.auxiliaryChainId = auxiliaryChainId;
+
+    // Suppressing lint error
+    // Refer: https://github.com/typescript-eslint/typescript-eslint/issues/636
+    /* eslint-disable @typescript-eslint/unbound-method */
+    this.reactTo = this.reactTo.bind(this);
+    this.update = this.update.bind(this);
+  }
+
+
+  /**
+   * This method react on changes in auxiliary chain models.
+   * @param auxiliaryChains List of auxiliary chains model
+   */
+  public async update(auxiliaryChains: AuxiliaryChain[]): Promise<void> {
+    const proveGatewayPromises = auxiliaryChains
+      .filter(ac => ac.chainId === this.auxiliaryChainId)
+      .map(
+        async (auxiliaryChain): Promise<{
+          success: boolean;
+          transactionHash: string;
+          message: string;
+        }> => (
+          auxiliaryChain.lastOriginBlockHeight
+            ? this.reactTo(auxiliaryChain.lastOriginBlockHeight)
+            : Promise.reject(new Error('Last anchored origin block height cannot be undefined.'))),
+      );
+
+    await Promise.all(proveGatewayPromises);
+  }
+
+  /**
+   * This method performs prove gateway transaction on auxiliary chain.
+   * This throws if auxiliary chain details doesn't exist.
+   *
+   * @param blockHeight Block height at which anchor state root happens.
+   *
+   * @return Return a promise that resolves to object which tell about success or failure.
+   */
+  public async reactTo(
```

Comment:
Before stake request service was merged, I finished development of prove gateway service. At that time `update` method interface was not known and defined. After stake request service was merged this `update` interface came. I don't want to redesign prove gateway service again so I kept `reactTo` method(which is now proveGateway method) and used it inside `update` method. It saves me time and effort. 

We can do this clean up later in time as I have already iterated on this service many times.

1. It's not needed but keeping it for now
2. Done
3. I have removed it.

Reply:
I see. If that's  big change, sure let's postpone it for future. However, to be sure that reactTo is not used from outside, I would still like to see it private.

Reply:
Making it private is the big change, due to test changes.

Reply:
I see. As a compromised solution here, I would propose to add a documentation, that it's public only because of testing and it should not be used as it's not part of public interface. Agree?

Reply:
Added comment.